libevmasm/RuleList.h

```diff
@@ -697,6 +699,28 @@ std::vector<SimplificationRule<Pattern>> evmRuleList(
 			[]() -> Pattern { return Instruction::SELFBALANCE; }, false
 		});
 
+	// @returns i if _x == 2**i
+	auto binaryLogarithm = [](Word const& _x) -> std::optional<Word> {
+		for (size_t i = 0; i < Pattern::WordSize; i++)
+			if (_x == (Word(1) << i))
+				return Word(i);
+		return {};
+	};
+	if (_evmVersion.hasBitwiseShifting())
+	{
+		rules.emplace_back(
+			Builtins::EXP(2, X),
+			[=]() -> Pattern { return Builtins::SHL(X, 1); },
+			false
+		);
+		rules.emplace_back(
+			Builtins::EXP(A, X),
+			[=]() -> Pattern { return Builtins::SHL(Builtins::MUL(X, *binaryLogarithm(A.d())), 1); },
+			false,
+			[=]() { return binaryLogarithm(A.d()).has_value() && *binaryLogarithm(A.d()) > 1; }
```

Comment:
Are `exp(0,x)` and `exp(1,x)` okay to be left unoptimised?

Reply:
Added them.